### PR TITLE
[Admin] Remove avatar

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

**When admin doesn't have avatar:**
Screenshot 2019-07-25 at 14 47 42

**When avatar is added:**
Screenshot 2019-07-24 at 09 22 50

**After removing avatar:**
Screenshot 2019-07-25 at 14 47 42

### CODE REVIEW

File: features/user/managing_administrators/removing_administartor_avatar.feature [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+    Background:
+        Given I am logged in as an administrator
+        And this administrator has the "troll.jpg" image as avatar
+
+    @ui
+    Scenario: Removing an administrator avatar
+        Given I am editing my details
+        When I remove the avatar image
+        And I should not see the "troll.jpg" avatar image in the top bar next to my name
```
Comment (GSadee) [inline features/user/managing_administrators/removing_administartor_avatar.feature]:
```suggestion
        Then I should not see the "troll.jpg" avatar image in the top bar next to my name
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorsContext.php [hunk 2]
```diff
@@ -333,6 +342,17 @@ public function iShouldSeeTheAvatarImageInTheTopBarNextToMyName(string $avatar):
         Assert::true($this->topBarElement->hasAvatarInMainBar($avatar));
     }
 
+    /**
+     * @Given /^I should not see the "([^"]*)" avatar image in the top bar next to my name$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorsContext.php]:
```suggestion
     * @Then I should not see the :avatar avatar image in the top bar next to my name
```

File: features/user/managing_administrators/removing_administartor_avatar.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    I want to remove an administrator avatar of an account
+
+    Background:
+        Given I am logged in as an administrator
+        And this administrator has the "troll.jpg" image as avatar
+
+    @ui
+    Scenario: Removing an administrator avatar
+        Given I am editing my details
+        When I remove the avatar image
```
Comment (Zales0123) [inline features/user/managing_administrators/removing_administartor_avatar.feature]: Maybe just `When I remove the avatar`?

File: features/user/managing_administrators/removing_administartor_avatar.feature (hunk 1 above)
Comment (Zales0123) [inline features/user/managing_administrators/removing_administartor_avatar.feature]: `Then`

Comment (Zales0123) [reply features/user/managing_administrators/removing_administartor_avatar.feature]: And what about the avatar in admin form?

File: src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+use Sylius\Component\Core\Model\AvatarImage;
+use Sylius\Component\Core\Repository\AvatarRepositoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+
+final class AvatarController
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php]:
```suggestion
final class RemoveAvatarAction
```

File: src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        RepositoryInterface $avatarRepository,
+        EngineInterface $templatingEngine,
+        RouterInterface $router
+    ) {
+        $this->avatarRepository = $avatarRepository;
+        $this->templatingEngine = $templatingEngine;
+        $this->router = $router;
+    }
+
+    public function removeAvatarAction(Request $request, string $id): Response
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php]:
```suggestion
    public function __invoke(Request $request, string $id): Response
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorsContext.php (hunk 2 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorsContext.php]: Could be written with turnip `@Then I should not see the :avatar avatar image in the top bar next to my name`

File: src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        /** @var AvatarImage $avatar */
+        $avatar = $this->avatarRepository->findOneByOwner($id);
+
+        if(null !== $avatar) {
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php]:
```suggestion
        if (null !== $avatar) {
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorsContext.php [hunk 7]
```diff
@@ -333,6 +342,17 @@ public function iShouldSeeTheAvatarImageInTheTopBarNextToMyName(string $avatar):
...
+     */
+    public function iShouldNotSeeTheAvatarImageInTheTopBarNextToMyName(string $avatar): void
+    {
+        $avatarPath = $this->sharedStorage->get($avatar);
+
+        Assert::true(!$this->topBarElement->hasAvatarInMainBar($avatarPath));
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorsContext.php]: `Assert::true(!` -> `Assert::false(`

File: src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            $this->avatarRepository->remove($avatar);
+        }
+
+        $url = $this->router->generate('sylius_admin_admin_user_update', ['id' => $id]);
+
+        return new RedirectResponse($url);
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php]: Could be merged into one line with the previous one

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml [hunk 9]
```diff
@@ -16,3 +16,10 @@ sylius_admin_admin_user:
             index:
                 icon: lock
     type: sylius.resource
+
+sylius_admin_admin_user_remove_avatar:
+    path: /user/{id}/edit
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml]:
```suggestion
    path: /users/{id}/remove-avatar
```

File: src/Sylius/Behat/Page/Admin/Administrator/UpdatePage.php [hunk 10]
```diff
@@ -53,6 +53,11 @@ public function hasAvatar(string $avatarPath): bool
         return strpos($srcPath, $avatarPath) !== false;
     }
 
+    public function removeAvatar(): void
+    {
+        $this->getDocument()->find('css', '#delete-button')->click();
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Administrator/UpdatePage.php]: `$this->getDocument()->clickLink('Delete')`?

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml [hunk 11]
```diff
@@ -16,3 +16,10 @@ sylius_admin_admin_user:
...
+    methods: [GET, PUT]
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml]: Do you need `GET` action here?

File: src/Sylius/Behat/Page/Admin/Administrator/UpdatePageInterface.php [hunk 12]
```diff
@@ -28,4 +28,6 @@ public function changePassword(string $password): void;
     public function changeLocale(string $localeCode): void;
 
     public function hasAvatar(string $avatarPath): bool;
+
+    public function removeAvatar(): void;
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/Administrator/UpdatePageInterface.php]: Move it before `hasAvatar`

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml [hunk 13]
```diff
@@ -16,3 +16,10 @@ sylius_admin_admin_user:
...
+    defaults:
+        id: $id
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml]: Is this needed?

File: src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php (hunk 6 above)
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php]: Missing space after `if`

File: src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig [hunk 14]
```diff
@@ -26,6 +26,7 @@
                 <label>{{ 'sylius.ui.avatar'|trans }}</label>
                 {{ form_row(form.avatar, {'label': false}) }}
             </div>
+            <a href="{{ path('sylius_admin_admin_user_remove_avatar', {'id': app.request.attributes.get('id')}) }}" id="delete-button" class="ui icon red labeled button"><i class="icon trash"></i> {{ 'sylius.ui.delete'|trans }}</a>
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig]: Too long line imo

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public function findOneByOwner(string $id): ?AvatarImage
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.owner = :user')
+            ->andWhere('user.id = :id')
+            ->setParameter('id', $id)
+            ->getQuery()
+            ->getResult()
+            ;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]: Wrong indentation

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml (hunk 11 above)
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml]: Does it need a `GET` method?

File: src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php (hunk 5 above)
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Controller/AvatarController.php]: `string $id` is not needed, we can get it from `$request` object

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Sylius\Bundle\CoreBundle\Doctrine\ORM;
+
+use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+use Sylius\Component\Core\Model\AvatarImage;
+use Sylius\Component\Core\Repository\AvatarRepositoryInterface;
+
+final class AvatarRepository extends EntityRepository implements AvatarRepositoryInterface
+{
+    public function findOneByOwner(string $id): ?AvatarImage
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]:
```suggestion
    public function findOneByOwnerId(string $ownerId): ?ImageInterface
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.owner = :user')
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]:
```suggestion
            ->andWhere('o.owner = :ownerId')
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            ->andWhere('user.id = :id')
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]: Not needed

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+            ->setParameter('id', $id)
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]:
```suggestion
            ->setParameter('ownerId', $ownerId)
```

File: src/Sylius/Bundle/CoreBundle/Form/Type/User/AvatarImageType.php [hunk 20]
```diff
@@ -14,6 +14,7 @@
 namespace Sylius\Bundle\CoreBundle\Form\Type\User;
 
 use Sylius\Bundle\CoreBundle\Form\Type\ImageType;
+use Symfony\Component\Form\Extension\Core\Type\ButtonType;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Form/Type/User/AvatarImageType.php]: 😢

File: src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig (hunk 14 above)
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig]: Shouldn't this be a form with a PUT method? With a link, we're using GET to update the administrator, which is not appropriate

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Doctrine\ORM;
+
+use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+use Sylius\Component\Core\Model\AvatarImage;
+use Sylius\Component\Core\Repository\AvatarRepositoryInterface;
+
+final class AvatarRepository extends EntityRepository implements AvatarRepositoryInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]:
```suggestion
final class AvatarImageRepository extends EntityRepository implements AvatarImageRepositoryInterface
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig (hunk 14 above)
Comment (Zales0123) [reply src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig]: Btw, this button probably should not be displayed when admin has no avatar set

File: src/Sylius/Component/Core/Repository/AvatarRepositoryInterface.php [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\Repository;
+
+use Sylius\Component\Core\Model\AvatarImage;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+interface AvatarRepositoryInterface extends RepositoryInterface
```
Comment (GSadee) [inline src/Sylius/Component/Core/Repository/AvatarRepositoryInterface.php]:
```suggestion
interface AvatarImageRepositoryInterface extends RepositoryInterface
```

File: src/Sylius/Bundle/CoreBundle/Form/Type/User/AvatarImageType.php (hunk 20 above)
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Form/Type/User/AvatarImageType.php]: Not needed, I suppose

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php (hunk 15 above)
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarRepository.php]: Wrong indent

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarImageRepository.php [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+final class AvatarImageRepository extends EntityRepository implements AvatarImageRepositoryInterface
+{
+    public function findOneByOwnerId(string $ownerId): ?ImageInterface
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.owner = :ownerId')
+            ->setParameter('ownerId', $ownerId)
+            ->getQuery()
+            ->getOneOrNullResult();
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarImageRepository.php]: The colon should be moved to the next line

File: src/Sylius/Behat/Context/Setup/AdminUserContext.php [hunk 24]
```diff
@@ -85,6 +85,14 @@ public function thereIsAnAdministratorWithName($username)
         $this->sharedStorage->set('administrator', $adminUser);
     }
 
+    /**
+     * @Given /^(this administrator) has the "([^"]*)" image as avatar$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/AdminUserContext.php]: Why didn't you add this step to `iHaveTheImageAsMyAvatar`. It should work

File: src/Sylius/Bundle/AdminBundle/Action/RemoveAvatarAction.php [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+        AvatarImageRepositoryInterface $avatarRepository,
+        EngineInterface $templatingEngine,
+        RouterInterface $router
+    ) {
+        $this->avatarRepository = $avatarRepository;
+        $this->templatingEngine = $templatingEngine;
+        $this->router = $router;
+    }
+
+    public function __invoke(Request $request): Response
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Action/RemoveAvatarAction.php]: This action should be protected by the csrf token. We need to render csrf in the template, send it together with form and assert if the token is valid

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarImageRepository.php [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+final class AvatarImageRepository extends EntityRepository implements AvatarImageRepositoryInterface
+{
+    public function findOneByOwnerId(string $ownerId): ?ImageInterface
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.owner = :ownerId')
+            ->setParameter('ownerId', $ownerId)
+            ->getQuery()
+            ->getOneOrNullResult()
+            ;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarImageRepository.php]:
```suggestion
        ;
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarImageRepository.php [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+namespace Sylius\Bundle\CoreBundle\Doctrine\ORM;
+
+use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+use Sylius\Component\Core\Model\AvatarImage;
+use Sylius\Component\Core\Model\ImageInterface;
+use Sylius\Component\Core\Repository\AvatarImageRepositoryInterface;
+
+final class AvatarImageRepository extends EntityRepository implements AvatarImageRepositoryInterface
+{
+    public function findOneByOwnerId(string $ownerId): ?ImageInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AvatarImageRepository.php]:
```suggestion
    public function findOneByOwnerId(AdminUserInterface $owner): ?ImageInterface
```

Passing the whole class would work better as a contract for me

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/Configuration.php [hunk 28]
```diff
@@ -86,6 +88,7 @@ private function addResourcesSection(ArrayNodeDefinition $node): void
                                     ->addDefaultsIfNotSet()
                                     ->children()
                                         ->scalarNode('model')->defaultValue(AvatarImage::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('repository')->defaultValue(AvatarImageRepository::class)->cannotBeEmpty()->end()
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/Configuration.php]: We cannot declare repositories like [here](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/ProductBundle/Resources/config/services/integrations/doctrine/orm.xml)

File: src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig [hunk 29]
```diff
@@ -26,6 +26,11 @@
                 <label>{{ 'sylius.ui.avatar'|trans }}</label>
                 {{ form_row(form.avatar, {'label': false}) }}
             </div>
+            {% if event_prefix == 'sylius.admin.admin_user.update' %}
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/views/AdminUser/_form.html.twig]: :( Do we want to show this part only on update action? This solution won't work if we rename route

**Comment by lchrusciel** [comment]
Thank you, Tomasz! :tada:

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml [hunk 30]
```diff
@@ -16,3 +16,9 @@ sylius_admin_admin_user:
             index:
                 icon: lock
     type: sylius.resource
+
+sylius_admin_admin_user_remove_avatar:
+    path: /users/{id}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/admin_user.yml]:
```
path: /users/{id}/remove-avatar
```

**Pull request merged**